### PR TITLE
Python: Fix agent option merge to support dict-defined tools

### LINKED ISSUE #4303: Python: Agent option merge drops dict-defined tools
### Description
`_merge_options` drops dict-defined tools when merging default and runtime options.

The merge logic de-duplicates tools by `getattr(tool, "name", None)`. For dict tools, that value is `None`, so runtime dict tools are incorrectly treated as duplicates and discarded.

### Code Sample
```python
from agent_framework._agents import _merge_options

base = {
    "tools": [
        {"type": "function", "function": {"name": "tool_a"}},
    ]
}
override = {
    "tools": [
        {"type": "function", "function": {"name": "tool_b"}},
    ]
}

merged = _merge_options(base, override)
print(merged["tools"])
# Actual: only tool_a
# Expected: tool_a and tool_b
```

### Error Messages / Stack Traces
No exception is raised; the runtime tool is silently dropped.

### Package Versions
agent-framework-core: source checkout at commit `02ba27493`

### Python Version
Python 3.13.5

### Additional Context
This impacts hosted/dict-style tool definitions passed via options.

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

When tools are defined as raw dictionaries (e.g. `{"type": "function", "function": {"name": "..."}}`), the option merge logic used `getattr(t, "name", None)` which always returned `None` for dicts, causing all dict-defined tools to be silently dropped as false duplicates.

Fixes #4303

### Description

The root cause was that `_merge_options` assumed tools were always objects with a `.name` attribute. Dict-defined tools store their name nested under `tool["function"]["name"]`, so `getattr` never found it and every tool resolved to `None`, collapsing them during deduplication. The fix introduces a `_get_tool_name` helper that handles both object-style and dict-style tool definitions, and uses it in the deduplication logic. Tests cover dict-only, mixed, and deduplication scenarios.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [X] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

---

> Note: PR autogenerated by moonbox3's agent

### CODE REVIEW

File: python/packages/core/agent_framework/_agents.py [hunk 1]
```diff
@@ -97,8 +107,8 @@ def _merge_options(base: dict[str, Any], override: dict[str, Any]) -> dict[str,
             continue
         if key == "tools" and result.get("tools"):
             # Combine tool lists, avoiding duplicates by name
-            existing_names = {getattr(t, "name", None) for t in result["tools"]}
-            unique_new = [t for t in value if getattr(t, "name", None) not in existing_names]
+            existing_names = {_get_tool_name(t) for t in result["tools"]}
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_agents.py]: Minor reliability note (pre-existing, not introduced here): if `_get_tool_name` returns `None` for multiple tools (e.g., malformed dicts missing a `function.name`), they will all collide in `existing_names` and only the first nameless tool will be kept. Consider excluding `None` from the set to avoid silently dropping tools.

```suggestion
            existing_names = {_get_tool_name(t) for t in result["tools"]} - {None}
```

File: python/packages/core/tests/core/test_agents.py [hunk 2]
```diff
@@ -880,6 +880,71 @@ def __init__(self, name):
...
+        ]
+    }
+
+    result = _merge_options(base, override)
+
+    assert len(result["tools"]) == 2
+    names = [_get_tool_name(t) for t in result["tools"]]
+    assert "tool_a" in names
+    assert "tool_b" in names
+
```
Comment (moonbox3) [inline python/packages/core/tests/core/test_agents.py:946]: This test verifies that mixed object+dict tools are *combined*, but doesn't check the deduplication case — where both types share the same name. Add a test where `base` has `MockTool("tool_a")` and `override` has a dict tool also named `"tool_a"`, then assert `len(result["tools"]) == 1`.

```suggestion

def test_merge_options_mixed_tools_deduplicates():
    """Test _merge_options deduplicates across object and dict tools with the same name."""

    class MockTool:
        def __init__(self, name):
            self.name = name

    base = {"tools": [MockTool("tool_a")]}
    override = {
        "tools": [
            {"type": "function", "function": {"name": "tool_a"}},
            {"type": "function", "function": {"name": "tool_b"}},
        ]
    }

    result = _merge_options(base, override)

    assert len(result["tools"]) == 2
    names = [_get_tool_name(t) for t in result["tools"]]
    assert names.count("tool_a") == 1
    assert "tool_b" in names

```

File: python/packages/core/tests/core/test_agents.py [hunk 3]
```diff
@@ -880,6 +880,71 @@ def __init__(self, name):
     assert "tool2" in tool_names
```
Comment (moonbox3) [inline python/packages/core/tests/core/test_agents.py:880]: Since `_get_tool_name` is imported and is a public-enough helper to be tested directly, add unit tests covering its edge cases (malformed dict, missing 'function' key, non-dict 'function' value, object without 'name').

```suggestion
def test_get_tool_name_from_object():
    class MockTool:
        def __init__(self, name):
            self.name = name

    assert _get_tool_name(MockTool("my_tool")) == "my_tool"


def test_get_tool_name_from_dict():
    tool_def = {"type": "function", "function": {"name": "my_tool"}}
    assert _get_tool_name(tool_def) == "my_tool"


def test_get_tool_name_missing_function_key():
    assert _get_tool_name({"type": "function"}) is None


def test_get_tool_name_function_not_dict():
    assert _get_tool_name({"type": "function", "function": "not_a_dict"}) is None


def test_get_tool_name_no_name_attr():
    assert _get_tool_name(object()) is None


```

File: python/packages/core/agent_framework/_agents.py (hunk 1 above)
Comment (moonbox3) [reply python/packages/core/agent_framework/_agents.py]: Fixed in `e32d6c66`.

File: python/packages/core/tests/core/test_agents.py (hunk 2 above)
Comment (moonbox3) [reply python/packages/core/tests/core/test_agents.py:946]: Fixed in `e32d6c66`.

File: python/packages/core/tests/core/test_agents.py (hunk 3 above)
Comment (moonbox3) [reply python/packages/core/tests/core/test_agents.py:880]: Fixed in `e32d6c66`.

File: python/packages/core/tests/core/test_agents.py [hunk 4]
```diff
@@ -880,6 +880,152 @@ def __init__(self, name):
...
+    class MockTool:
+        def __init__(self, name):
+            self.name = name
+
+    assert _get_tool_name(MockTool("my_tool")) == "my_tool"
+
+
 def test_merge_options_logit_bias_merged():
     """Test _merge_options merges logit_bias dicts."""
     base = {"logit_bias": {"token1": 1.0}}
```
Comment (moonbox3) [inline python/packages/core/tests/core/test_agents.py:1031]: Good coverage of the core fix, but consider adding a combined test that exercises both dedup and the None-exclusion in one call. For example, base has a named tool 'tool_a' and a nameless tool, override has a duplicate 'tool_a' and another nameless tool — assert result has 1 'tool_a' and 2 nameless tools (total 3).

**Review by moonbox3 — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 3 | **Confidence:** 92%

### ✓ Correctness
> The change correctly prevents false deduplication of tools whose names cannot be extracted (i.e., `_get_tool_name` returns `None`). By removing `None` from the `existing_names` set, nameless tools in the override list are no longer incorrectly matched against nameless tools in the base list. The accompanying tests thoroughly cover edge cases for `_get_tool_name` and the deduplication logic. No correctness issues found.

### ✓ Security Reliability
> The diff is a small, well-scoped reliability fix that prevents tools with no extractable name (where `_get_tool_name` returns `None`) from being falsely deduplicated during option merging. The fix correctly removes `None` from the dedup set so that nameless tools from the override are always preserved. Accompanying tests thoroughly cover edge cases of `_get_tool_name` and the deduplication behavior. No security or reliability concerns found.

### ✓ Test Coverage
> The diff adds a small but important fix to exclude None from the deduplication set in _merge_options, and provides thorough test coverage for both the fix and the _get_tool_name helper. The core bug scenario (nameless tools falsely deduplicated) is directly tested, and edge cases for _get_tool_name are well covered. One minor gap: there is no test combining nameless and named tools in the same merge to verify both behaviors work simultaneously.

### Suggestions
- Consider adding a combined scenario test where base contains both a nameless tool and a named tool, and override contains both a nameless tool and a duplicate-named tool. This would verify that nameless tools are preserved while named duplicates are still correctly deduplicated in a single merge call.

---
*Automated review by moonbox3's agents*

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/core/agent_framework</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py">_agents.py</a></td><td>338</td><td>43</td><td>87%</td><td><a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L435">435</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L439">439</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L491">491</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L856">856</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L892">892</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L908">908</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L984-L987">984&ndash;987</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1048-L1050">1048&ndash;1050</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1171">1171</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1187">1187</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1189">1189</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1202">1202</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1208">1208</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1244">1244</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1246">1246</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1255-L1260">1255&ndash;1260</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1265">1265</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1267">1267</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1273-L1274">1273&ndash;1274</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1281">1281</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1283-L1284">1283&ndash;1284</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1292-L1293">1292&ndash;1293</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1296-L1298">1296&ndash;1298</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1306-L1307">1306&ndash;1307</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1309">1309</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1314">1314</a>, <a href="https://github.com/microsoft/agent-framework/blob/e32d6c669f70d789d784ab4aea62b88af0125fb2/packages/core/agent_framework/_agents.py#L1316">1316</a></td></tr><tr><td><b>TOTAL</b></td><td><b>22181</b></td><td><b>2762</b></td><td><b>87%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 4684 | 247 :zzz: | 0 :x: | 0 :fire: | 1m 19s :stopwatch: |
